govcd/api_token.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2021 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"github.com/vmware/go-vcloud-director/v2/util"
+)
+
+// SetApiToken behaves similarly to SetToken, with the difference that it will
+// return full information about the bearer token, so that the caller can make decisions about token expiration
+func (vcdCli *VCDClient) SetApiToken(org, apiToken string) (*types.ApiTokenRefresh, error) {
+	usableApiToken, err := vcdCli.GetBearerTokenFromApiToken(org, apiToken)
+	if err != nil {
+		return nil, err
+	}
+	err = vcdCli.SetToken(org, BearerTokenHeader, usableApiToken.AccessToken)
+	if err != nil {
+		return nil, err
+	}
+	return usableApiToken, nil
+}
+
+// GetBearerTokenFromApiToken receives an API token and retrieves a bearer token
+// using the refresh token operation.
+func (vcdCli *VCDClient) GetBearerTokenFromApiToken(org, token string) (*types.ApiTokenRefresh, error) {
+	if vcdCli.Client.APIVCDMaxVersionIs("< 36.1") {
+		version, err := vcdCli.Client.GetVcdFullVersion()
+		if err == nil {
+			return nil, fmt.Errorf("minimum version for API token is 10.3.1 - Version detected: %s", version.Version)
+		}
+		// If we can't get the VCD version, we return API version info
+		return nil, fmt.Errorf("minimum API version for API token is 36.1 - Version detected: %s", vcdCli.Client.APIVersion)
+	}
+	var userDef string
+	urlStr := strings.Replace(vcdCli.Client.VCDHREF.String(), "/api", "", 1)
+	if strings.EqualFold(org, "system") {
+		userDef = "provider"
+	} else {
+		userDef = fmt.Sprintf("tenant/%s", org)
+	}
+	reqUrl := fmt.Sprintf("%s/oauth/%s/token", urlStr, userDef)
+	reqHref, err := url.ParseRequestURI(reqUrl)
+	if err != nil {
+		return nil, fmt.Errorf("error getting request URL from %s : %s", reqUrl, err)
+	}
+
+	options := map[string]string{
+		"grant_type":    "refresh_token",
+		"refresh_token": token,
+	}
+	req := vcdCli.Client.NewRequest(options, http.MethodPost, *reqHref, nil)
+	req.Header.Add("Accept", "application/*;version=36.1")
+
+	resp, err := vcdCli.Client.Http.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	var body []byte
+	var tokenDef types.ApiTokenRefresh
+	if resp.Body != nil {
+		body, err = ioutil.ReadAll(resp.Body)
+	}
+	responseData := "[" + strings.Repeat("*", 10) + "]"
```

Comment:
Please add a comment why are we doing this.

Reply:
This is an obfuscation of sensitive data for logging. 
Comment added